File: slp/backend-dotnet/Features/Queue/RedisConnectionFactory.cs
using Microsoft.Extensions.Logging;
using StackExchange.Redis;

namespace backend_dotnet.Features.Queue;

public class RedisConnectionFactory
{
    private readonly Lazy<IConnectionMultiplexer?> _lazyConnection;
    private readonly ILogger<RedisConnectionFactory> _logger;

    public RedisConnectionFactory(string connectionString, ILogger<RedisConnectionFactory> logger)
    {
        _logger = logger;
        _lazyConnection = new Lazy<IConnectionMultiplexer?>(() =>
        {
            if (string.IsNullOrEmpty(connectionString))
            {
                _logger.LogWarning("Redis connection string is missing or empty.");
                return null;
            }

            try
            {
                return ConnectionMultiplexer.Connect(connectionString);
            }
            catch (Exception ex)
            {
                _logger.LogWarning(ex, "Failed to connect to Redis. Queue service will operate in degraded mode.");
                return null;
            }
        });
    }

    public IConnectionMultiplexer? Connection => _lazyConnection.Value;
}

File: slp/backend-nestjs/src/modules/queue/redis-connection.factory.ts
import { Injectable, Logger } from '@nestjs/common';
import Redis from 'ioredis';

@Injectable()
export class RedisConnectionFactory {
  private readonly logger = new Logger(RedisConnectionFactory.name);
  private readonly redis: Redis | null;

  constructor() {
    const connectionString = process.env.REDIS_URL;
    if (!connectionString) {
      this.logger.warn('Redis connection string is missing.');
      this.redis = null;
      return;
    }

    try {
      this.redis = new Redis(connectionString);
      this.logger.log('Redis connected');
    } catch (err) {
      this.logger.warn('Failed to connect to Redis', err);
      this.redis = null;
    }
  }

  getConnection(): Redis | null {
    return this.redis;
  }
}